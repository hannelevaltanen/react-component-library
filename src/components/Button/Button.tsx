import React from "react";
import "./button.css";

export interface ButtonProps {
  /**
   * Is this the principal call to action on the page?
   */
  primary?: boolean;
  /**
   * What background color to use
   */
  backgroundColor?: string;
  /**
   * How large should the button be?
   */
  size?: "small" | "medium" | "large";
  /**
   * Button contents
   */
  children: React.ReactNode;
  /**
   * Optional click handler
   */
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

/**
 * Primary UI component for user interaction
 */
const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      primary = true,
      backgroundColor,
      size = "medium",
      onClick,
      children,
    }: ButtonProps,
    ref: React.Ref<HTMLButtonElement>
  ) => {
    const mode = primary
      ? "storybook-button--primary"
      : "storybook-button--secondary";
    return (
      <button
        ref={ref}
        type="button"
        className={["storybook-button", `storybook-button--${size}`, mode].join(
          " "
        )}
        style={backgroundColor ? { backgroundColor } : {}}
        onClick={onClick}
      >
        <span>{children}</span>
      </button>
    );
  }
);

export default Button;
