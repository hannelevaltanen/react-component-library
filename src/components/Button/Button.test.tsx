import React from "react";
import { render } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";

import Button from "./Button";

expect.extend(toHaveNoViolations);

describe("<Button /> spec", () => {
  it("renders the component", () => {
    const { asFragment } = render(<Button>My Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should not have basic accessibility issues", async () => {
    const { container } = render(<Button>My Button</Button>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
