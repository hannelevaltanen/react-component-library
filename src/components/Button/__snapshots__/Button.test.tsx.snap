// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> spec renders the component 1`] = `
<DocumentFragment>
  <button
    class="storybook-button storybook-button--medium storybook-button--primary"
    type="button"
  >
    <span>
      My Button
    </span>
  </button>
</DocumentFragment>
`;
